### ISSUE TITLE
code

### ISSUE THREAD

**Issue by viciou0sneed5** [viciou0sneed5/himanshu#1]
import feedparser
from newspaper import Article
import spacy
import nltk
from nltk.corpus import stopwords
import re
from transformers import pipeline
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
from flask import Flask, jsonify, request

# Initialize NLP tools
print("Loading NLP models...")
nlp = spacy.load("en_core_web_sm")
nltk.download("stopwords")
stop_words = set(stopwords.words("english"))

# Load NLP models
classifier = pipeline("zero-shot-classification", model="facebook/bart-large-mnli")
summarizer = pipeline("summarization", model="facebook/bart-large-cnn")
fake_news_detector = pipeline("text-classification", model="microsoft/deberta-v3-base")
analyzer = SentimentIntensityAnalyzer()

CATEGORIES = ["Politics", "Technology", "Sports", "Entertainment", "Business", "Health"]

def fetch_google_news_rss():
    print("Fetching news from Google RSS...")
    rss_url = "https://news.google.com/rss"
    feed = feedparser.parse(rss_url)
    articles = []
    for entry in feed.entries[:10]:
        try:
            url = entry.link
            print(f"Processing article: {url}")
            article = Article(url)
            article.download()
            article.parse()
            articles.append({
                "title": article.title,
                "content": article.text,
                "url": url,
                "published": entry.published
            })
        except Exception as e:
            print(f"Error processing {url}: {e}")
    print("News fetching complete.")
    return articles

def clean_text(text):
    print("Cleaning text...")
    text = re.sub(r"\s+", " ", text).lower()
    doc = nlp(text)
    return " ".join([token.lemma_ for token in doc if token.text not in stop_words and token.is_alpha])

def classify_news(text):
    print("Classifying news...")
    result = classifier(text, CATEGORIES)
    return result["labels"][0]

def summarize_article(text):
    print("Summarizing article...")
    summary = summarizer(text, max_length=100, min_length=30, do_sample=False)
    return summary[0]["summary_text"]

def analyze_sentiment(text):
    print("Analyzing sentiment...")
    score = analyzer.polarity_scores(text)
    return "Positive" if score["compound"] > 0 else "Negative" if score["compound"] < 0 else "Neutral"

def detect_fake_news(text):
    print("Detecting fake news...")
    result = fake_news_detector(text)
    return result[0]["label"]

def recommend_articles(user_history, all_articles):
    print("Generating recommendations...")
    corpus = [user_history] + [article["content"] for article in all_articles]
    vectorizer = TfidfVectorizer().fit_transform(corpus)
    similarities = cosine_similarity(vectorizer[0], vectorizer[1:])
    recommended_idx = similarities.argsort()[0][-3:]
    return [all_articles[i]["title"] for i in recommended_idx]

app = Flask(__name__)

@app.route("/fetch_news", methods=["GET"])
def fetch_news():
    print("Fetching and processing news...")
    articles = fetch_google_news_rss()
    for article in articles:
        print(f"Processing article: {article['title']}")
        article["category"] = classify_news(article["content"])
        article["summary"] = summarize_article(article["content"])
        article["sentiment"] = analyze_sentiment(article["content"])
        article["fake_news"] = detect_fake_news(article["content"])
    print("News processing complete.")
    return jsonify(articles)

@app.route("/recommend", methods=["POST"])
def recommend():
    print("Processing recommendation request...")
    data = request.json
    user_history = data["user_history"]
    articles = data["articles"]
    recommendations = recommend_articles(user_history, articles)
    print("Recommendation processing complete.")
    return jsonify({"recommendations": recommendations})

if __name__ == "__main__":
    print("Starting Flask server...")
    app.run(debug=True, port=5000)
